Add a BLE_SupportedFeatures example

idea--list opened the pull request:
Main purpose of this new example is to list which optional BLE features are supported by a device and showcases how one can access BLE related API members. It also starts advertising of non-connectable type and thus implements listening to only 1 event, which helps to understand the concept of event listeners before overwhelming the user with 10 different event listeners in the first example. As such, this example is aimed for those who just begin programming BLE with Mbed OS.

paul-szczepanek-arm commented on BLE_SupportedFeatures/readme.md:
```diff
@@ -0,0 +1,22 @@
+
+This application tells you which optional BLE features are supported by your device. This info can be useful as product datasheets often do not go into these details.
+Knowing which optional features are supported might impact how exactly you implement BLE feature of your applications.
+This demo also showcases how to access BLE related API methods via your BLE object instance.
```
I think it would be enough for this example to init and show features, the advertising is already showcases in the dedicated example.

idea--list replied:
That is right, the advertising part of this example is not necessary. Just thought that no BLE examples make sense without any type of advertising at least.

paul-szczepanek-arm commented on BLE_SupportedFeatures/source/main.cpp:
```diff
@@ -0,0 +1,164 @@
+    SupportedFeatures(BLE &ble, events::EventQueue &event_queue) :
+        _ble(ble),
+        _event_queue(event_queue),
+        _service_uuid(GattService::UUID_DEVICE_INFORMATION_SERVICE),
+        _adv_data_builder(_adv_buffer)
```
we can skip these

paul-szczepanek-arm commented on BLE_SupportedFeatures/source/main.cpp:
```diff
@@ -0,0 +1,164 @@
+        /* display basic info and optional BLE features supported by our device */
+
+        printf("BLE Version %s\n", _ble.getVersion());
+        printf("Conn handle %d\n", ble::connection_handle_t(&_ble));
```
This case doesn't make sense. What was your intention here?

idea--list replied:
These can be removed, i just tried to fetch as much data about the BLE instance as possible.

paul-szczepanek-arm commented on BLE_SupportedFeatures/source/main.cpp:
```diff
@@ -0,0 +1,164 @@
+        printf("LE_POWER_CLASS feature supported %d\n", _ble.gap().isFeatureSupported(ble::controller_supported_features_t::LE_POWER_CLASS));
+
+        /* this allows us to capture events like onConnectionComplete() */
+        _ble.gap().setEventHandler(this);
+
+        start_advertising();
```
no need

paul-szczepanek-arm commented on BLE_SupportedFeatures/source/main.cpp:
```diff
@@ -0,0 +1,164 @@
+        _adv_data_builder.setLocalServiceList({&_service_uuid, 1});
+        _adv_data_builder.setName(DEVICE_NAME);
+
+
+        /* Set advertising parameters configuration to our BLE instance */
+
+        ble_error_t error = _ble.gap().setAdvertisingParameters(
+            ble::LEGACY_ADVERTISING_HANDLE,
+            adv_parameters
+        );
+
+        if (error) {
+            printf("_ble.gap().setAdvertisingParameters() failed\r\n");
+            return;
+        }
+
+        /* Set advertising payload configuration to our BLE instance */
+
+        error = _ble.gap().setAdvertisingPayload(
+            ble::LEGACY_ADVERTISING_HANDLE,
+            _adv_data_builder.getAdvertisingData()
+        );
+
+        if (error) {
+            printf("_ble.gap().setAdvertisingPayload() failed\r\n");
+            return;
+        }
+
+        /* Start advertising */
+
+        error = _ble.gap().startAdvertising(ble::LEGACY_ADVERTISING_HANDLE);
+
+        if (error) {
+            printf("_ble.gap().startAdvertising() failed\r\n");
+            return;
+        }
+
+        printf("Device is advertising\r\n");
+
+    }
```
this is a duplicate of the dedicated advertising demo, I think we can skip this

idea--list replied:
Yes, i started from that example, not from scratch.

paul-szczepanek-arm commented on BLE_SupportedFeatures/source/main.cpp:
```diff
@@ -0,0 +1,164 @@
+    BLE &_ble;
+    events::EventQueue &_event_queue;
+
+    UUID _service_uuid;
+
+    uint8_t _adv_buffer[ble::LEGACY_ADVERTISING_MAX_SIZE];
+    ble::AdvertisingDataBuilder _adv_data_builder;
```
we can skip these

paul-szczepanek-arm commented on BLE_SupportedFeatures/source/main.cpp:
```diff
@@ -0,0 +1,164 @@
+
+        printf("BLE Version %s\n", _ble.getVersion());
+        printf("Conn handle %d\n", ble::connection_handle_t(&_ble));
+        printf("LE_ENCRYPTION feature supported %d\n", _ble.gap().isFeatureSupported(ble::controller_supported_features_t::LE_ENCRYPTION));
```
Since we're only interested in supported feature maybe instead of:
```printf("feature supported: %d", isFeatureSupported())```
we could instead:
```
if (isFeatureSupported(X)) {
    printf("Feature X supported");
}
```

idea--list replied:
Actually i would prefer keeping it as i posted it. It is also good to see which features are not supported.

paul-szczepanek-arm replied:
OK
```suggestion
        printf("LE_ENCRYPTION - %s\n", _ble.gap().isFeatureSupported(ble::controller_supported_features_t::LE_ENCRYPTION)? "supported" : "not supported");
```

idea--list replied:
With strings it is indeed more user-friemndly

pan- replied:
I tend to agree. I would prefer to see what's supported as well as what is not supported.

paul-szczepanek-arm reviewed (changes requested):
Great idea. I don't think we need to add advertising to this one though. We already have a dedicated advertising example. I think it's valuable in itself to initialise and print features.

paul-szczepanek-arm commented on BLE_SupportedFeatures/mbed-os.lib:
```diff
@@ -0,0 +1 @@
+https://github.com/ARMmbed/mbed-os/#26606218ad9d1ee1c8781aa73774fd7ea3a7658e
```
```suggestion
https://github.com/ARMmbed/mbed-os/

```

idea--list replied:
That is right, i forked this repo yesterday, made the PR and apparently this fetched the hash of a specific release

paul-szczepanek-arm replied:
That is my oversight, I will make a note in the readme about what branch to target with PRs.

paul-szczepanek-arm commented:
PRs need to target the development branch. Master is only for releases.

paul-szczepanek-arm commented:
To rebase your branch:
`git pull --rebase upstream development`
assuming you have upstream defined, if not, first add it with:
`git remote add upstream https://github.com/ARMmbed/mbed-os-example-ble.git`

idea--list commented:
will give git cli a try, though until now i only used github.com directly...

idea--list commented:
I am on a windows10 machine with Git bash installed.
For some reason i can not clone the repo using git commands, so i downloaded&extracted the .zip.
Then inside git bash i changed to the folder where i extracted the repo, typed `git init`
then `git remote add upstream https://github.com/ARMmbed/mbed-os-example-ble.git` and also `git pull --rebase upstream development` executed as expected.
However `git push -f` results in: 

```
fatal: No configured push destination.
Either specify the URL from the command-line or configure a remote repository using

    git remote add <name> <url>

and then push using the remote name

    git push <name>
```
here i got stuck

idea--list commented:
@paul-szczepanek-arm 
Instead of fighting with git commands i installed "github desktop".
Cloned this repo successfully, then i under Branch->rebase current branch i tried to change to the development branch but it tells me the development branch is up to date with master and does not let me rebase.
![image](https://user-images.githubusercontent.com/5784071/110425701-0febc180-80a5-11eb-9240-ea41ea6d8288.png)

idea--list commented:
If i try to clone the development branch with github desktop, then it asks for my username&pw but does not let me clone the dev branch. Something might be missing.

paul-szczepanek-arm commented:
If you cloned it with github desktop you can open the command line (in the repository menu -> open terminal) and use the git commands.

The original problem you had is because you didn't clone the reop you didn't have origin set up. You add it the same way:
`git remote add origin https://github.com/idea--list/mbed-os-example-ble.git`

idea--list commented:
I think the easiest solution is that i delete this PR, fork the repo again and then i edit the development branch in my fork and make a new PR based on that.

paul-szczepanek-arm commented:
Rebasing is a very common procedure, it's worth learning it.

paul-szczepanek-arm commented:
Are you sure you checked out your own fork and not this repo?

idea--list commented:
> Are you sure you checked out your own fork and not this repo? I mean through the desktop version.
Maybe that was the issue, let's see...

idea--list commented:
Still no luck:
1. I cloned my own fork on my desktop.
2. Under repository menu -> open in command prompt i type `git remote add upstream https://github.com/ARMmbed/mbed-os-example-ble.git` this returns: error: remote upstream already exists.
3. Then i type `git pull --rebase upstream development` in github desktop a popup is shown telling me there is a conflict with the mbed-os.lib file in each example folder.
4. I edit all the conflicting files to contain nothing just: `https://github.com/ARMmbed/mbed-os/`
5. Doing so i may click the "continue rebase" button.
6. Then i needed to repeat steps 4 & 5.
7. Now it seems to OK.

What is next?

paul-szczepanek-arm commented:
github desktop already adds upstream, that's why the error. It's fine. Now you have a rebased branch on your local machine you just need to push but you have to do it on the command line:
`git push -f`
The -f is short for --force-push which overwrites the branch at origin (github) with your newly rebased version on your local machine.

idea--list commented:
The command prompt tells me it is done.
Is there anything else to do on my behalf?

paul-szczepanek-arm commented:
Yes, remove the advertising part from the example.

idea--list commented:
Thanks for leading me through my first rebasing process!
I will make the changes today on my local desktop and then force push once again.

paul-szczepanek-arm commented:
OK, thanks. Force push is only needed if the remote has a different history from yours (which is what happens if you rebase - you change the history/base of your commits), otherwise normal push is enough.

idea--list commented:
Made the changes, updated the readme, pushed to my fork on github.
Hope everything is fine this time.

paul-szczepanek-arm commented:
Thank you. Tomorrow I will merge this to a separate branch temporarily to clean up the history a bit and then open a PR for development. I'll add you as reviewer when I do that so you can approve any changes I make.
